src/Exceptionless.Core/Services/StackService.cs

```diff
@@ -0,0 +1,113 @@
+            if (String.IsNullOrEmpty(organizationId))
+                throw new ArgumentNullException(nameof(organizationId));
+            if (String.IsNullOrEmpty(projectId))
+                throw new ArgumentNullException(nameof(projectId));
+            if (String.IsNullOrEmpty(stackId))
+                throw new ArgumentNullException(nameof(stackId));
+            if (count == 0)
+                return;
+
+            await Task.WhenAll(
+                _cache.SetAddAsync(GetStackOccurrenceSetCacheKey(), Tuple.Create(organizationId, projectId, stackId)),
+                _cache.IncrementAsync(GetStackOccurrenceCountCacheKey(organizationId, projectId, stackId), count, _expireTimeout),
+                _cache.SetIfLowerAsync(GetStackOccurrenceMinDateCacheKey(organizationId, projectId, stackId), ConvertToUnixTimeStamp(minOccurrenceDateUtc), _expireTimeout),
+                _cache.SetIfHigherAsync(GetStackOccurrenceMaxDateCacheKey(organizationId, projectId, stackId), ConvertToUnixTimeStamp(maxOccurrenceDateUtc), _expireTimeout)
+            ).AnyContext();
+        }
+
+        public async Task SaveStackUsagesAsync(bool sendNotifications = true, CancellationToken cancellationToken = default) {
+            var occurrenceSetCacheKey = GetStackOccurrenceSetCacheKey();
+            var stackUsageSet = await _cache.GetSetAsync<Tuple<string, string, string>>(occurrenceSetCacheKey).AnyContext();
+            if (!stackUsageSet.HasValue || stackUsageSet.IsNull) return;
+            foreach (var tuple in stackUsageSet.Value) {
+                if (cancellationToken.IsCancellationRequested) break;
+
+                string organizationId = tuple.Item1, projectId = tuple.Item2, stackId = tuple.Item3;
+                string occurrenceCountCacheKey = GetStackOccurrenceCountCacheKey(organizationId, projectId, stackId),
+                    occurrenceMinDateCacheKey = GetStackOccurrenceMinDateCacheKey(organizationId, projectId, stackId),
+                    occurrenceMaxDateCacheKey = GetStackOccurrenceMaxDateCacheKey(organizationId, projectId, stackId);
+                var occurrenceCountTask = _cache.GetAsync<long>(occurrenceCountCacheKey, 0);
+                var occurrenceMinDateTask = _cache.GetAsync<double>(occurrenceMinDateCacheKey);
+                var occurrenceMaxDateTask = _cache.GetAsync<double>(occurrenceMaxDateCacheKey);
+
+                await Task.WhenAll(occurrenceCountTask, occurrenceMinDateTask, occurrenceMaxDateTask).AnyContext();
+                var occurrenceCount = (int)occurrenceCountTask.Result;
+                if (occurrenceCount == 0) continue;
+
+                await Task.WhenAll(
+                    _cache.DecrementAsync(occurrenceCountCacheKey, occurrenceCount, _expireTimeout),
+                    _cache.SetRemoveAsync(occurrenceSetCacheKey, tuple),
```

Comment:
I think this set remove task should be moved up to the previous when all on line 57 to further prevent a race condition.